src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/BinderEquivalence.cs

```diff
@@ -26,16 +28,27 @@ internal static class BinderEquivalence
         // it does not need to be very precise either
         private static int cachedBinderCount;
 
-        // it is unlikely to see a lot of contention on the binder cache.
-        // creating binders is not a very frequent operation.
-        // typically a dynamic operation in the source will create just one binder lazily when first executed.
-        private static readonly ConcurrentDictionary<ICSharpBinder, ICSharpBinder> binderEquivalenceCache =
-            new ConcurrentDictionary<ICSharpBinder, ICSharpBinder>(concurrencyLevel: 2, capacity: 32, new BinderEqualityComparer());
+        // Keep a separate cache per ALC to allow unloadability.
+        private static readonly ConditionalWeakTable<AssemblyLoadContext,
+            ConcurrentDictionary<ICSharpBinder, ICSharpBinder>> binderEquivalenceCache =
+            new ConditionalWeakTable<AssemblyLoadContext, ConcurrentDictionary<ICSharpBinder, ICSharpBinder>>();
 
-        internal static T TryGetExisting<T>(this T binder)
+        internal static T TryGetExisting<T>(this T binder, Type context)
             where T : ICSharpBinder
         {
-            var fromCache = binderEquivalenceCache.GetOrAdd(binder, binder);
+            var alc = AssemblyLoadContext.GetLoadContext(context.Assembly);
```

Comment:
This won't work well for non-collectible generic types instantiated over collectible types. E.g. consider how this is going to work for `List<MyCollectibleType>`. `context.Assembly` is going to be `System.Private.CoreLib`, so this type is going to be cached against the default non-unloadable load context that is not correct.

Reply:
Hmm, how about we just don't cache the binders if the context type `IsCollectible`?